### PR TITLE
Prusti Test: Hashmap

### PR DESCRIPTION
**Pull request by JM4ier**
an implementation of a hashmap based on [veribetrfs's linear hashmap](https://github.com/vmware-labs/verified-betrfs/blob/master/lib/DataStructures/LinearMutableMap.i.dfy).

verification is lacking at the moment, but a few things lead to unexpected errors.

@vakaras

### CODE REVIEW

**Comment by JM4ier** [comment]
@vakaras

File: prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[package]
+name = "veribetrfs"
+version = "0.1.0"
+edition = "2018"
+
+[dependencies]
+
+[workspace]
```
Comment (fpoli) [inline prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml]: Could you make sure that all files end with a newline? If you open the [files](https://github.com/viperproject/prusti-dev/pull/923/files) tab of the PR GitHub shows a red sign where the newline is missing.

File: prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[package]
+name = "veribetrfs"
+version = "0.1.0"
+edition = "2018"
+
+[dependencies]
```
Comment (fpoli) [inline prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml:6]:
```suggestion
[dependencies]
prusti-contracts = { path = "prusti-contracts" }
```

I expect the CI to fail if this is missing. See [here](https://github.com/viperproject/prusti-dev/blob/master/prusti-tests/tests/cargo_verify/failing_crate/Cargo.toml) for an example.

Comment (JM4ier) [reply prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml:6]: I added the dependency now.

Comment (JM4ier) [reply prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml:6]: After adding the dependency, running Prusti locally fails, as there is no dependency "prusti-contracts" in that path.
I tried adding symbolic links but it got messy really quickly, as the transitive dependencies of all the other "prusti-*" crates also are searched in that directory.

Is there an easy way to fix this?

Comment (fpoli) [reply prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml:6]: When running the tests (`./x.py test` to run all tests, `./x.py test -p prusti-tests --test cargotest` to run just the `cargo_verify` tests) it's the test suite that creates those symbolic links.

If you want to launch `cargo build/prusti` on your own to debug, it should be enough to create the symbolic links to these five folders: https://github.com/viperproject/prusti-dev/blob/dacab2d43007ed6eeb88d87d2a04b061f8684f91/prusti-tests/tests/cargotest.rs#L141-L145 Does that work?

Comment (JM4ier) [reply prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml:6]: yes, that works

File: prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "veribetrfs"
+version = "0.1.0"
+edition = "2018"
```
Comment (vakaras) [inline prusti-tests/tests/cargo_verify/veribetrfs/Cargo.toml:4]:
```suggestion
edition = "2021"
```

File: prusti-tests/tests/cargo_verify/veribetrfs/src/base/option.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
+use prusti_contracts::*;
+
+pub enum Opt<T> {
```
Comment (vakaras) [inline prusti-tests/tests/cargo_verify/veribetrfs/src/base/option.rs:3]: You can use the built-in Rust `Option` type by giving it specs as shown [here](https://github.com/viperproject/prusti-dev/blob/e51e0c21b5f0ec5ad7e7ca9b3b8eda322965c0cb/prusti-tests/tests/verify/pass/extern-spec/option.rs).

File: prusti-tests/tests/cargo_verify/veribetrfs/src/base/mod.rs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+pub mod option;
+pub mod vector;
+
+pub use option::Opt;
+pub use vector::Vector;
+
+pub fn replace<T>(dest: &mut T, src: T) -> T {
```
Comment (vakaras) [inline prusti-tests/tests/cargo_verify/veribetrfs/src/base/mod.rs:7]: You can use the `std::mem::replace` directly by giving it specs as shown [here](https://github.com/viperproject/prusti-dev/blob/e51e0c21b5f0ec5ad7e7ca9b3b8eda322965c0cb/prusti-tests/tests/verify/pass/extern-spec/swap.rs).

File: prusti-tests/tests/cargo_verify/veribetrfs/src/base/vector.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
+use prusti_contracts::*;
+
+pub struct Vector<T>(Vec<T>);
+
+impl<T> Vector<T> {
```
Comment (vakaras) [inline prusti-tests/tests/cargo_verify/veribetrfs/src/base/vector.rs:5]: You can find examples showing how to add specs to Rust's `Vec` [here](https://github.com/viperproject/prusti-dev/tree/e51e0c21b5f0ec5ad7e7ca9b3b8eda322965c0cb/prusti-tests/tests/verify/pass/extern-spec). However, I am not sure whether `index` and `index_mut` are properly supported.

**Comment by vakaras** [comment]
Add `/home/data/eth/rust/prusti-dev2/prusti-tests/tests/cargotest.rs`:
```rust
#[cargo_test]
fn test_veribetrfs() {
    test_local_project("veribetrfs");
}
```

File: prusti-tests/tests/cargo_verify/veribetrfs/src/base/vector.rs (hunk 6 above)
Comment (Aurel300) [reply prusti-tests/tests/cargo_verify/veribetrfs/src/base/vector.rs:5]: *Might* be supported after #882. `index` is actually a method of the `Index` trait though, so the extern spec will be something like `#[extern_spec] impl<T> Index<usize> for Vec<T> { ... }`

**Pull request merged**